### ISSUE TITLE
App using Powershell.SDK fails to run, when building as SingleFile

### ISSUE THREAD

**Issue by RobinGFT** [dotnet/runtime#102193]
### Description

I have a .NET 8 Application executing a lot of Powershell commands, but when I build it as a SingleFile it throws an exception.
Issue is already documented in the powershell repo: [#13540](https://github.com/PowerShell/PowerShell/issues/13540)

### Reproduction Steps

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net8.0-windows</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.PowerShell.SDK" Version="7.4.2"  />
  </ItemGroup>
</Project>
```

```
using System;
using System.Management.Automation;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            String cmd = "New-Item -Force -Type Directory -Path '%userprofile%\test123'";
             PowerShell ps1 = PowerShell.Create().AddScript(cmd);
             ps1.Invoke()
        }
    }
}
```

`dotnet publish -r win-x64 --self-contained false /p:PublishSingleFile=true`

### Expected behavior

Publish and run successfully

### Actual behavior

Publish works, but application doesn' t run. At the first execution of a powershell script, following exception is thrown: 

```
ERROR: 2024-05-14T12:33:51.8318820Z - Value cannot be null. (Parameter 'path1')
   at System.ArgumentNullException.Throw(String paramName)
   at System.IO.Path.Combine(String path1, String path2)
   at System.Management.Automation.PSSnapInReader.ReadEnginePSSnapIns()
   at System.Management.Automation.Runspaces.InitialSessionState.CreateDefault()
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace(PSHost host)
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace()
   at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
   at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TOutput](IEnumerable input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke()
```

### Configuration

* .NET 8
* Win10-x64
* PowerShell.SDK 7.4.2

**Comment by MichalStrehovsky** [comment]
ArgumentNullException from `Path.Combine` that happens when compiled in single file is a telltale sign of the app calling `Assembly.Location`, passing it to some path API to get the directory and then combining that with something else. There's a lot of references to `Assembly.Location` in PowerShell, not sure which one this is, but I'm 99% certain the problem is that.

[Assembly.Location](https://learn.microsoft.com/en-us/dotnet/core/deploying/single-file/overview?tabs=cli#api-incompatibility) is documented to return an empty string in single file apps. The fix for this is very likely going to be in the PowerShell codebase. PowerShell codebase would have to first set `<EnableSingleFileAnalyzer>true</EnableSingleFileAnalyzer>` property in the project to enable analyzer that flags all problematic API and the revise the behavior for flagged APIs.

**Comment by RobinGFT** [comment]
> ArgumentNullException from `Path.Combine` that happens when compiled in single file is a telltale sign of the app calling `Assembly.Location`, passing it to some path API to get the directory and then combining that with something else. There's a lot of references to `Assembly.Location` in PowerShell, not sure which one this is, but I'm 99% certain the problem is that.
> 
> [Assembly.Location](https://learn.microsoft.com/en-us/dotnet/core/deploying/single-file/overview?tabs=cli#api-incompatibility) is documented to return an empty string in single file apps. The fix for this is very likely going to be in the PowerShell codebase. PowerShell codebase would have to first set `<EnableSingleFileAnalyzer>true</EnableSingleFileAnalyzer>` property in the project to enable analyzer that flags all problematic API and the revise the behavior for flagged APIs.

Hi Michael - thank you for your immediate answer!

In the PowerShell repo there is already an issue, which still isn't fixed, but closed: [PowerShell/PowerShell #13540](https://github.com/PowerShell/PowerShell/issues/13540)
Maybe I will open another issue there with linking to this one. 
Is there any way, I can exclude NuGET packages from SingleFile publish?

**Comment by jkotas** [comment]
> Is there any way, I can exclude NuGET packages from SingleFile publish?

You can exclude individul assemblies from being emdedded: https://learn.microsoft.com/en-us/dotnet/core/deploying/single-file/overview?tabs=cli#exclude-files-from-being-embedded . You will likely need to play a lot of whac-a-mole to find all places.

> Maybe I will open another issue there with linking to this one.

Right, this would have to be fixed in Powershell. The discussion in https://github.com/PowerShell/PowerShell/issues/13540 suggests that there are a lot of places to fix.
